lib/widgets/editor/editor_toolbar.dart

```diff
@@ -0,0 +1,469 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+
+import '../../hooks/logged_in_action.dart';
+import '../../l10n/l10n.dart';
+import '../../markdown_formatter.dart';
+import '../../resources/links.dart';
+import '../../url_launcher.dart';
+import '../../util/async_store_listener.dart';
+import '../../util/extensions/api.dart';
+import '../../util/extensions/spaced.dart';
+import '../../util/files.dart';
+import '../../util/mobx_provider.dart';
+import '../../util/observer_consumers.dart';
+import '../../util/text_lines_iterator.dart';
+import 'editor_picking_dialog.dart';
+import 'editor_toolbar_store.dart';
+
+class _Reformat {
+  final String text;
+  final int selectionBeginningShift;
+  final int selectionEndingShift;
+  _Reformat({
+    required this.text,
+    this.selectionBeginningShift = 0,
+    this.selectionEndingShift = 0,
+  });
+}
+
+enum HeaderLevel {
+  h1(1),
+  h2(2),
+  h3(3),
+  h4(4),
+  h5(5),
+  h6(6);
+
+  const HeaderLevel(this.value);
+  final int value;
+}
+
+class EditorToolbar extends HookWidget {
+  final TextEditingController controller;
+  final String instanceHost;
+  final EditorToolbarStore store;
+  final FocusNode editorFocusNode;
+  static const _height = 50.0;
+
+  EditorToolbar({
+    required this.controller,
+    required this.instanceHost,
+    required this.editorFocusNode,
+  }) : store = EditorToolbarStore(instanceHost);
+
+  @override
+  Widget build(BuildContext context) {
+    final visible = useState(editorFocusNode.hasFocus);
+    useEffect(() {
+      editorFocusNode.addListener(() {
+        visible.value = editorFocusNode.hasFocus;
+      });
+
+      return;
+    }, [editorFocusNode]);
+
+    return MobxProvider(
+      create: (context) => store,
```

Comment:
Why like this? Why not created here? I understand one extra `Builder` might seem annoying, but it would be really beneficial to have a clear source of this store